### PR TITLE
Add "IsExternal" switch to Job Sender SDK

### PR DESCRIPTION
**Pull request by jonfortescue**
Adds the ability to send jobs anonymously using an authenticated API for storage requests which effectively ports the "IsExternal" option from BuildTools.

Now tested and working!

### CODE REVIEW

File: src/Microsoft.DotNet.Helix/Sdk/HelixTask.cs [hunk 1]
```diff
@@ -20,11 +20,18 @@ public abstract class HelixTask : Task
         /// </summary>
         public string AccessToken { get; set; }
 
+        /// <summary>
+        /// A boolean value determining whether this job should be considered "external" for creation purposes
+        /// </summary>
+        public bool IsExternal { get; set; } = false;
```
Comment (alexperovich) [inline src/Microsoft.DotNet.Helix/Sdk/HelixTask.cs]: Can you put this property on the SendHelixJob task? It doesn't make sense on the base class. Just fill both properties, and the derived classes can pick which one to use.

File: src/Microsoft.DotNet.Helix/Sdk/HelixTask.cs [hunk 2]
```diff
@@ -20,11 +20,18 @@ public abstract class HelixTask : Task
...
 
-        private IHelixApi GetHelixApi()
+        protected IHelixApi AnonymousApi { get; private set; }
+
+        private IHelixApi GetHelixApi(bool requestAnonymous = false)
         {
-            if (string.IsNullOrEmpty(AccessToken))
+            if (string.IsNullOrEmpty(AccessToken) || requestAnonymous)
             {
                 Log.LogMessage(MessageImportance.Low, "No AccessToken provided, using anonymous access to helix api.");
```
Comment (alexperovich) [inline src/Microsoft.DotNet.Helix/Sdk/HelixTask.cs]: We probably don't want to log this if `requestAnonymous` is true.

File: src/Microsoft.DotNet.Helix/Sdk/HelixTask.cs [hunk 3]
```diff
@@ -38,6 +45,10 @@ public sealed override bool Execute()
         {
             try
             {
+                if (IsExternal)
+                {
+                    AnonymousApi = GetHelixApi(requestAnonymous: true);
```
Comment (alexperovich) [inline src/Microsoft.DotNet.Helix/Sdk/HelixTask.cs]: I think its easier here to remove the `requestAnonymous` parameter and just do
```suggestion
                    AnonymousApi = ApiFactory.GetAnonymous(BaseUri)
```

File: src/Microsoft.DotNet.Helix/Sdk/HelixTask.cs (hunk 1 above)
Comment (jonfortescue) [reply src/Microsoft.DotNet.Helix/Sdk/HelixTask.cs]: okay sounds fine

File: src/Microsoft.DotNet.Helix/Sdk/HelixTask.cs (hunk 2 above)
Comment (jonfortescue) [reply src/Microsoft.DotNet.Helix/Sdk/HelixTask.cs]: fair

File: src/Microsoft.DotNet.Helix/Sdk/HelixTask.cs (hunk 3 above)
Comment (jonfortescue) [reply src/Microsoft.DotNet.Helix/Sdk/HelixTask.cs]: fair

**Comment by jonfortescue** [comment]
@alexperovich Do these changes work for you?

**Comment by chcosta** [comment]
>  I genuinely don't understand how this keeps happening

Is it possible that your fork of arcade has an extra commit in its history?  One thing I do when I see weirdness like this is to go to my fork and...

```
> git checkout master
> git pull upstream master
> git reset upstream/master
> git push origin +master
```

Assuming upstream == "github.com/dotnet/arcade" and origin == "github.com/jonfortescue/arcade"

**Comment by jonfortescue** [comment]
@chcosta Possible. It's fixed for now at least.

**Comment by jonfortescue** [comment]
This has been fully tested and is working as intended. Merging.

File: src/Microsoft.DotNet.Helix/Sdk/SendHelixJob.cs [hunk 4]
```diff
@@ -123,12 +134,35 @@ protected override async Task ExecuteCore()
...
                     .WithBuild(Build)
                     .WithTargetQueue(TargetQueue)
                     .WithMaxRetryCount(MaxRetryCount);
+                Log.LogMessage($"Initialized job definition with source '{Source}', type '{Type}', build number '{Build}', and target queue '{TargetQueue}'");
+
+                if (IsExternal)
+                {
+                    if (string.IsNullOrEmpty(Creator))
+                    {
+                        Log.LogError("The Creator property was left unspecified for an external job. Please set the Creator property or use set IsExternal to false.");
```
Comment (MattGal) [inline src/Microsoft.DotNet.Helix/Sdk/SendHelixJob.cs]: - extraneous "use" before "set IsExternal"
- Isn't the property surfaced to the user AnonymousApi not IsExternal?

Comment (jonfortescue) [reply src/Microsoft.DotNet.Helix/Sdk/SendHelixJob.cs]: @MattGal good catch. I'll fix that in #1503. The property is IsExternal, though.

**Pull request merged**